File: simple-stack-example-fragments/src/main/java/com/zhuinden/simplestackdemoexamplefragments/application/BaseKey.java
package com.zhuinden.simplestackdemoexamplefragments.application;

import android.os.Bundle;
import android.support.v4.app.Fragment;

import com.zhuinden.simplestackdemoexamplefragments.util.BaseFragment;

/**
 * Created by Owner on 2017. 02. 03..
 */

public abstract class BaseKey
        implements Key {
    @Override
    public final Fragment newFragment() {
        Fragment fragment = createFragment();
        Bundle bundle = fragment.getArguments();
        if(bundle == null) {
            bundle = new Bundle();
        }
        bundle.putParcelable(BaseFragment.KEY_TAG, this);
        fragment.setArguments(bundle);
        return fragment;
    }

    protected abstract Fragment createFragment();

    @Override
    public String getFragmentTag() {
        return toString();
    }
}

File: simple-stack-example-mvp-fragments/src/main/java/com/zhuinden/simplestackdemoexamplefragments/application/BaseKey.kt
package com.zhuinden.simplestackdemoexamplefragments.application

import android.os.Bundle
import android.support.v4.app.Fragment

import com.zhuinden.simplestackdemoexamplefragments.util.BaseFragment

/**
 * Created by Zhuinden on 2018. 08. 20.
 */

abstract class BaseKey : Key {
    override fun newFragment(): Fragment = createFragment().also { fragment ->
        fragment.arguments = (fragment.arguments ?: Bundle()).also { bundle ->
            bundle.putParcelable(BaseFragment.KEY_TAG, this)
        }
    }

    protected abstract fun createFragment(): Fragment

    override val fragmentTag: String
        get() = toString()
}
